### PR TITLE
net: arp: Timeout too long ARP request

### LINKED ISSUE #8019: ARP: should drop any packet pended when timeout
Zephyr send ARP packet first before ICMP if the ip address doesn't cached in ip tables,  always pended the ICMP packet and resend pended ICMP packet when it got ARP reply.

But if the ARP packet doesn't have a  reply because host doesn't exist, zephyr will always hold on the ICMP packet pended, doesn't free ! Even if doesn't have a ARP reply there always has a ARP item in arp tables associate that ip address but mac address is zero.

Next is the log from zephyr's console shell, I try ping some nonexistent hosts:

```
net> ping 192.168.3.10                                                                           
Sent a ping to 192.168.3.10                                                                      
Ping timeout                                                                                     
net> arp                                                                                         
     Interface  Link              Address                                                        
[ 0] 0x2000dd60 00:00:00:00:00:00 192.168.3.10                                                   
net> mem                                                                                         
Fragment length 128 bytes                                                                        
Network buffer pools:                                                                            
Address         Total   Avail   Name                                                             
0x2000dc5c      32      32      RX                                                               
0x2000dc78      32      31      TX                                                               
0x2000dcf4      32      32      RX DATA (rx_bufs)                                                
0x2000dd1c      32      31      TX DATA (tx_bufs)                                                
net> ping 192.168.3.11                                                                           
Sent a ping to 192.168.3.11                                                                      
Ping timeout                                                                                     
net> arp                                                                                         
     Interface  Link              Address                                                        
[ 0] 0x2000dd60 00:00:00:00:00:00 192.168.3.10                                                   
[ 1] 0x2000dd60 00:00:00:00:00:00 192.168.3.11                                                   
net> mem                                                                                         
Fragment length 128 bytes                                                                        
Network buffer pools:                                                                            
Address         Total   Avail   Name                                                             
0x2000dc5c      32      32      RX                                                               
0x2000dc78      32      30      TX                                                               
0x2000dcf4      32      32      RX DATA (rx_bufs)                                                
0x2000dd1c      32      30      TX DATA (tx_bufs)                                                
net> ping 192.168.3.12                                                                           
Sent a ping to 192.168.3.12                                                                      
Ping timeout                                                                                     
net> arp                                                                                         
     Interface  Link              Address                                                        
[ 0] 0x2000dd60 00:00:00:00:00:00 192.168.3.10                                                   
[ 1] 0x2000dd60 00:00:00:00:00:00 192.168.3.11                                                   
[ 2] 0x2000dd60 00:00:00:00:00:00 192.168.3.12                                                   
net> mem                                                                                         
Fragment length 128 bytes                                                                        
Network buffer pools:                                                                            
Address         Total   Avail   Name                                                             
0x2000dc5c      32      32      RX                                                               
0x2000dc78      32      29      TX                                                               
0x2000dcf4      32      32      RX DATA (rx_bufs)                                                
0x2000dd1c      32      29      TX DATA (tx_bufs)                                                
net> ping 192.168.3.13                                                                           
Sent a ping to 192.168.3.13                                                                      
Ping timeout                                                                                     
net> arp                                                                                         
     Interface  Link              Address                                                        
[ 0] 0x2000dd60 00:00:00:00:00:00 192.168.3.10                                                   
[ 1] 0x2000dd60 00:00:00:00:00:00 192.168.3.11                                                   
[ 2] 0x2000dd60 00:00:00:00:00:00 192.168.3.12                                                   
[ 3] 0x2000dd60 00:00:00:00:00:00 192.168.3.13                                                   
net> mem                                                                                         
Fragment length 128 bytes                                                                        
Network buffer pools:                                                                            
Address         Total   Avail   Name                                                             
0x2000dc5c      32      32      RX                                                               
0x2000dc78      32      28      TX                                                               
0x2000dcf4      32      32      RX DATA (rx_bufs)                                                
0x2000dd1c      32      28      TX DATA (tx_bufs)                                                
net> ping 192.168.3.14                                                                           
Sent a ping to 192.168.3.14                                                                      
Ping timeout                                                                                     
net> arp                                                                                         
     Interface  Link              Address                                                        
[ 0] 0x2000dd60 00:00:00:00:00:00 192.168.3.10                                                   
[ 1] 0x2000dd60 00:00:00:00:00:00 192.168.3.11                                                   
[ 2] 0x2000dd60 00:00:00:00:00:00 192.168.3.12                                                   
[ 3] 0x2000dd60 00:00:00:00:00:00 192.168.3.13                                                   
[ 4] 0x2000dd60 00:00:00:00:00:00 192.168.3.14                                                   
net> mem                                                                                         
Fragment length 128 bytes                                                                        
Network buffer pools:                                                                            
Address         Total   Avail   Name                                                             
0x2000dc5c      32      32      RX                                                               
0x2000dc78      32      27      TX                                                               
0x2000dcf4      32      32      RX DATA (rx_bufs)                                                
0x2000dd1c      32      27      TX DATA (tx_bufs)                                                
net> 
```

There should resolve those problem:

1. do NOT occupy one item in arp tables if that ARP doesn't has a reply
2. free net packet pended before when ARP timeout

### PR DESCRIPTION
**Pull request by jukkar**
ARP requests were not timeout.

Fixes #8019

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

### CODE REVIEW

**Comment by zhaozhongchen** [comment]
A simple test: 

```
shell> select net 
net> ping 192.168.3.10
Sent a ping to 192.168.3.10
Ping timeout
net> arp
ARP cache is empty.
net> ping 192.168.3.11
Sent a ping to 192.168.3.11                                                                      
Ping timeout                                                                                     
net> arp                                                                                         
ARP cache is empty.                                                                              
net> mem                                                                                         
Fragment length 128 bytes                                                                        
Network buffer pools:                                                                            
Address         Total   Avail   Name                                                             
0x2000de9c      32      32      RX                                                               
0x2000deb8      32      32      TX                                                               
0x2000df34      32      32      RX DATA (rx_bufs)                                                
0x2000df5c      32      32      TX DATA (tx_bufs)                                                
net> 
```

File: subsys/net/ip/l2/ethernet/arp.c [hunk 1]
```diff
@@ -194,6 +198,22 @@ static inline struct net_pkt *prepare_arp(struct net_if *iface,
...
+static void arp_request_timeout(struct k_work *work)
+{
+	/* This means that the ARP failed. */
+	struct arp_entry *entry = CONTAINER_OF(work,
+					       struct arp_entry,
+					       arp_request_timer);
+
+	if (entry->pending) {
+		NET_DBG("Releasing pending pkt %p (ref %d)", entry->pending,
+			entry->pending->ref - 1);
```
[the remark this answers is not in the document]
Comment (jukkar) [reply subsys/net/ip/l2/ethernet/arp.c]: Actually the comment is in right place, the function should only be called if the ARP request timeout. I just noticed that the current patch was missing one timeout cancel call which I will add.

[the remark this answers is not in the document]
Comment (zhaozhongchen) [reply subsys/net/ip/l2/ethernet/arp.c]: seems you haven't cancel delayed work when got arp reply

[the remark this answers is not in the document]
Comment (jukkar) [reply subsys/net/ip/l2/ethernet/arp.c]: As I said, the new version of this PR will cancel the work after receiving arp reply. See line 385 few lines below.

**Comment by jukkar** [comment]
The bug is low priority so setting the target release to 1.13

File: subsys/net/ip/l2/ethernet/arp.c [hunk 2]
```diff
@@ -361,6 +381,8 @@ static inline void arp_update(struct net_if *iface,
 				&NET_ETH_HDR(arp_table[i].pending)->dst.addr;
 
 			send_pending(iface, &arp_table[i].pending);
+
+			k_delayed_work_cancel(&arp_table[i].arp_request_timer);
```
Comment (pfalcon) [inline subsys/net/ip/l2/ethernet/arp.c]: Shouldn't we cancel timeout ASAP, instead of at the end of processing?

**Review by pfalcon — COMMENTED** [review]
A question to consider.

**Comment by jukkar** [comment]
Fixed the merge conflict. Please review.

**Review by pfalcon — APPROVED** [review]

**Pull request merged**